### PR TITLE
RADTTS Training

### PR DESCRIPTION
**Pull request by zachwe**

### CODE REVIEW

**Comment by sjkoelle** [comment]
Looks like the precommit black hasnt been running as we would hope recently so there are a fair number of cosmetically altered files here too

File: uberduck_ml_dev/utils/utils.py [hunk 1]
```diff
@@ -136,16 +136,23 @@ def dynamic_range_decompression(x, C=1):
...
     return x
 
 
+def to_gpu_radtts(batch):
+
+    output = {k: to_gpu(v) for k, v in batch}
+    return output
+
+
 def get_mask_from_lengths(lengths: torch.Tensor, max_len: int = 0):
```
Comment (sjkoelle) [inline uberduck_ml_dev/utils/utils.py:156]: dedupe

File: ray_dataset_exp.ipynb
Comment (sjkoelle) [inline ray_dataset_exp.ipynb:1]: remove

File: settings.ini [hunk 2]
```diff
@@ -25,7 +25,7 @@ license = apache2
 status = 2
 
 # Optional. Same format as setuptools requirements.  Torch version seems to effect random number generator (not 100% certain).
-requirements = Cython pytest phonemizer inflect wandb librosa==0.8.1 matplotlib nltk>=3.6.5 numpy>=1.20 csvw clldutils pandas pydub scipy scikit-learn soundfile tensorboardX torch>=1.9.0 torchaudio>=0.9.0 unidecode seaborn mdutils wordcloud wordfreq Pillow einops g2p_en@git+https://github.com/uberduck-ai/g2p emoji text-unidecode gdown pre-commit hyperpyyaml@git+https://github.com/speechbrain/HyperPyYAML speechbrain@git+https://github.com/speechbrain/speechbrain
```
Comment (sjkoelle) [inline settings.ini:28]: make sure speechbrain embedding dependency is removed

File: uberduck_ml_dev/data/audio_processing.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+import torch
+import numpy as np
+from scipy.signal import get_window
+from librosa.filters import mel as librosa_mel_fn
+import librosa.util as librosa_util
+
+
+def window_sumsquare(
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/audio_processing.py:28]: this is redundant with code in utils/utils

File: uberduck_ml_dev/data/audio_processing.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+import torch.nn.functional as F
+from torch.autograd import Variable
+from scipy.signal import get_window
+from librosa.util import pad_center, tiny
+
+
+class STFT(torch.nn.Module):
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/audio_processing.py:214]: redundant with code in models/common

File: uberduck_ml_dev/data/hifigan.py [hunk 5]
```diff
@@ -47,47 +47,86 @@ def spectral_de_normalize_torch(magnitudes):
...
+    with open(a.input_validation_file, "r", encoding="utf-8") as fi:
+        validation_files = [
+            os.path.join(a.input_wavs_dir, x.split("|")[0])
+            for x in fi.read().split("\n")
+            if len(x) > 0
+        ]
     return training_files, validation_files
 
 
 class MelDataset(torch.utils.data.Dataset):
```
Comment (sjkoelle) [inline uberduck_ml_dev/data/hifigan.py:110]: move to uberduck_ml_dev/data

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 6]
```diff
@@ -0,0 +1,1480 @@
...
+        for audiopath in audiopaths:
+            audio_data = read(audiopath)[1]
+            datum = torch.FloatTensor(audio_data).unsqueeze(-1).t().cuda()
+            # datum = torch.FloatTensor(audio_data).unsqueeze(-1).t()
+            emb = self.model(datum)
+            emb = emb.cpu().detach().numpy()
+            yield {"audio_embedding": emb}
+
+
+class DataCollate:
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:95]: move to /data

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 7]
```diff
@@ -0,0 +1,1480 @@
...
+            "attn_prior": attn_prior_padded,
+            "f0": f0_padded,
+            "p_voiced": p_voiced_padded,
+            "voiced_mask": voiced_mask_padded,
+            "energy_avg": energy_avg_padded,
+            "audio_embedding": audio_embedding_padded,
+        }
+
+
+class Data(torch.utils.data.Dataset):
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:211]: move to /data

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 8]
```diff
@@ -0,0 +1,1480 @@
...
+        # mel_path = data['mel_path']
+        f0, voiced_mask, p_voiced = torch.load(f0_path)
+        f0 = self.f0_normalize(f0)
+        if self.distance_tx_unvoiced:
+            mask = f0 <= 0.0
+            distance_map = np.log(distance_transform(mask))
+            distance_map[distance_map <= 0] = 0.0
+            f0 = f0 - distance_map
+
+        # if data['lmdb_key'] is not None:
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:523]: remove, but figure out: what was the point of this code in original repo?

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 9]
```diff
@@ -0,0 +1,1480 @@
...
+    #         self.cache_data_lmdb.get(prior_path.encode("ascii"))
+    #     )
+    if os.path.exists(prior_path):
+        attn_prior = torch.load(prior_path)
+    else:
+        attn_prior = beta_binomial_prior_distribution(
+            n_tokens, n_frames, scaling_factor=1.0  # 0.05
+        )
+        torch.save(attn_prior, prior_path)
+    # else:
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:616]: remove

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 10]
```diff
@@ -0,0 +1,1480 @@
...
+def get_attention_prior(n_tokens, n_frames):
+    # cache the entire attn_prior by filename
+    # if self.use_attn_prior_masking:
+    filename = "{}_{}".format(n_tokens, n_frames)
+    betabinom_cache_path = "yourmom"
+    if not os.path.exists(betabinom_cache_path):
+        os.makedirs(betabinom_cache_path, exist_ok=False)
+    prior_path = os.path.join(betabinom_cache_path, filename)
+    prior_path += "_prior.pth"
+    # if self.lmdb_cache_path != "":
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:605]: see above

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 11]
```diff
@@ -0,0 +1,1480 @@
...
+    return x
+
+
+def get_energy_average(mel):
+    energy_avg = mel.mean(0)
+    energy_avg = energy_avg_normalize(energy_avg)
+    return energy_avg
+
+
+def beta_binomial_prior_distribution(phoneme_count, mel_count, scaling_factor=0.05):
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:583]: seems like it should get its own folder, priors? precompute?  or just utils

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 12]
```diff
@@ -0,0 +1,1480 @@
...
+        torch.save(attn_prior, prior_path)
+    # else:
+    # attn_prior = torch.ones(n_frames, n_tokens)  # all ones baseline
+
+    return attn_prior
+
+
+# NOTE (Sam): looks like this was not used in successful training runs
+def f0_normalize(x, f0_min):
+    # if self.use_log_f0:
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:624]: okay to leave but maybe we should actually try this

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 13]
```diff
@@ -0,0 +1,1480 @@
...
+
+    return x
+
+
+def get_speaker_id(speaker):
+
+    return torch.LongTensor([speaker])
+
+
+def get_text(text):
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:637]: to text

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 14]
```diff
@@ -0,0 +1,1480 @@
...
+    text = tp.encode_text(text)
+    text = torch.LongTensor(text)
+    return text
+
+
+def get_shuffle_indices(levels):
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:643]: not used but very cool so find a folder

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 15]
```diff
@@ -0,0 +1,1480 @@
...
+    levels_unique = np.unique(levels)
+    output_indices = np.zeros(len(levels), dtype=int)
+    for level in levels_unique:
+        indices = np.where(levels == level)[0]
+        new_indices = np.random.permutation(indices)
+        output_indices[indices] = new_indices
+    return output_indices
+
+
+def ray_df_preprocessing(df):
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:654]: data / ray?

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 16]
```diff
@@ -0,0 +1,1480 @@
...
+    transcripts = df.transcript.tolist()
+    audio_bytes_list = df.audio_bytes.tolist()
+    speaker_ids = df.speaker_id.tolist()
+    paths = df.path.tolist()
+    f0_paths = df.f0_path.tolist()
+    audio_embeddings = df.audio_embedding.tolist()
+    # NOTE (Sam): I'm great at naming things.
+    # shuffle_indices = np.load('/usr/src/app/radtts/30shuffle_sdfixed_indices.pt.npy')[:100]
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:662]: remove

Comment (sjkoelle) [reply uberduck_ml_dev/exec/train_radtts_with_ray.py:662]: just clean this whole section

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 17]
```diff
@@ -0,0 +1,1480 @@
...
+        prior_path += "_prior.pth"
+        attn_prior = torch.load(prior_path)
+        # attn_prior = get_attention_prior(text_sequence.shape[0], mel.shape[1])
+        speaker_id = get_speaker_id(speaker_id)
+        # datum = torch.FloatTensor(audio_norm).unsqueeze(-1).t().cuda()
+        # audio_embedding = torch.load(emb_path)
+        audio_embedding = torch.FloatTensor(audio_embedding)
+        # audio_embeddings = audio_encoder(datum)
+        # audio_embeddings = None
+        # NOTE (Sam): might be faster to return dictionary arrays of batched inputs instead of list
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:706]: started to try this - got to dictionary of dictionaries i think (ray doesnt have built in padding for time series) but didnt evaluate speed.

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 18]
```diff
@@ -0,0 +1,1480 @@
...
+                "audiopath": None,
+                "audio_embedding": audio_embedding,
+            }
+        )
+        # print(datetime.now(), 'end')
+
+    return collate_input
+
+
+def get_ray_dataset():
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:726]: see above

File: uberduck_ml_dev/exec/train_radtts_with_ray.py [hunk 19]
```diff
@@ -0,0 +1,1480 @@
...
+        },
+        filepath,
+    )
+
+
+# NOTE (Sam): not necessary for ray dataset
+collate_fn = DataCollate()
+
+
+def _train_step(
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_radtts_with_ray.py:1007]: to trainer

File: uberduck_ml_dev/models/components/encoders/resnet_speaker_encoder.py [hunk 20]
```diff
@@ -36,12 +35,13 @@
     "db_level": -27.0,
 }
 
+RESNET_SE_MODEL_URL = "https://uberduck-models-us-west-2.s3.us-west-2.amazonaws.com/coqui/resnet_se.pth.tar"
```
Comment (sjkoelle) [inline uberduck_ml_dev/models/components/encoders/resnet_speaker_encoder.py:38]: remove

File: uberduck_ml_dev/exec/train_tacotron2_with_ray.py [hunk 21]
```diff
@@ -0,0 +1,330 @@
+from io import BytesIO
```
Comment (sjkoelle) [inline uberduck_ml_dev/exec/train_tacotron2_with_ray.py:1]: remove

File: uberduck_ml_dev/exec/split_train_val.py
Comment (sjkoelle) [inline uberduck_ml_dev/exec/split_train_val.py:1]: dont remove
